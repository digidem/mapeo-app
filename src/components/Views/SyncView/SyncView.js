// @flow
import React from 'react';
import {
  TouchableOpacity,
  TouchableWithoutFeedback,
  Text,
  View,
  FlatList,
  Dimensions,
  StyleSheet,
  NetInfo
} from 'react-native';
import { NavigationActions, withNavigationFocus } from 'react-navigation';
import WifiIcon from 'react-native-vector-icons/MaterialIcons';
import SyncHeader from './SyncHeader';
import DeviceCell from './DeviceCell';
import SyncedModal from '../../Base/SyncedModal/SyncedModal';
import type { Device } from '../../../types/device';
import { MAPEO_BLUE, MEDIUM_BLUE } from '../../../lib/styles';
import I18n from 'react-native-i18n';

type Props = {
  navigation: NavigationActions
};

export type StateProps = {
  devices: Device[],
  selectedDevice?: Device,
  syncAnnounced: boolean,
  syncedModalVisible: boolean
};

export type DispatchProps = {
  announceSync: () => void,
  listDevices: () => void,
  selectDevice: (device?: Device) => void,
  toggleDeviceSelect: (device: Device) => void,
  updateDeviceSync: (device: Device) => void,
  showSyncedModal: () => void,
  hideSyncedModal: () => void
};

type State = {
  wifi: boolean
};

if (I18n) {
  I18n.fallbacks = true;
  I18n.translations = {
    en: require('../../../translations/en'),
    es: require('../../../translations/es')
  };
}

class SyncView extends React.Component<
  Props & StateProps & DispatchProps,
  State
> {
  state = { wifi: false };

  componentDidMount() {
    const { announceSync, listDevices } = this.props;

    announceSync();
    listDevices();
    NetInfo.getConnectionInfo().then(connectionInfo => {
      if (connectionInfo.type === 'wifi') {
        this.setState({ wifi: true });
      }
    });
    NetInfo.addEventListener('connectionChange', this.handleConnectionChange);
  }

  componentWillUnmount() {
    NetInfo.removeEventListener(
      'connectionChange',
      this.handleConnectionChange
    );
  }

  shouldComponentUpdate(
    nextProps: Props & StateProps & DispatchProps,
    nextState: State
  ) {
    if (nextProps.navigation.isFocused()) {
      return nextProps !== this.props || nextState !== this.state;
    }

    return false;
  }

  handleConnectionChange = connectionInfo => {
    if (connectionInfo.type === 'wifi') {
      this.setState({ wifi: true });
    } else {
      this.setState({ wifi: false });
    }
  };

  render() {
    const {
      devices,
      navigation,
      selectDevice,
      selectedDevice,
      toggleDeviceSelect,
      updateDeviceSync,
      hideSyncedModal,
      showSyncedModal,
      syncedModalVisible
    } = this.props;
    const { wifi } = this.state;

    let syncStopped = false;
    const noDevices = devices.length === 0 || !wifi;
    let headerDeviceText = noDevices
      ? I18n.t('sync.none')
      : I18n.t('sync.available');
    if (selectedDevice) {
      switch (selectedDevice.syncStatus) {
        case 'requested':
          headerDeviceText = I18n.t('sync.initiated');
          break;
        case 'syncing':
          headerDeviceText = I18n.t('sync.progress');
          break;
        case 'stopped':
          headerDeviceText = I18n.t('sync.stopped');
          syncStopped = true;
          break;
        case 'completed':
          headerDeviceText = 'Sync completed';
          break;
        default:
          headerDeviceText = I18n.t('sync.selected');
      }
    }

    const keyExtractor = (item, index) => item.id;

    const handleDevicePress = item => {
      const syncInProgress =
        item.syncStatus === 'requested' || item.syncStatus === 'syncing';
      const startSync = () => {
        updateDeviceSync({
          ...item,
          selected: true,
          syncStatus: 'replication-started'
        });
        setTimeout(() => {
          updateDeviceSync({
            ...item,
            selected: true,
            syncStatus: 'replication-progress'
          });
        }, 1000);
        setTimeout(() => {
          updateDeviceSync({
            ...item,
            selected: true,
            syncStatus: 'replication-complete'
          });
          toggleDeviceSelect(item);
          showSyncedModal();
        }, 4000);
      };
      if (selectedDevice) {
        updateDeviceSync({
          ...selectedDevice,
          selected: false,
          syncStatus: 'replication-stopped'
        });

        if (selectedDevice.id !== item.id) {
          startSync();
        }
      } else {
        startSync();
      }
    };
    const renderItem = ({ item }) => (
      <DeviceCell
        device={item}
        onPress={handleDevicePress}
        selectedDevice={selectedDevice}
        showSyncedModal={showSyncedModal}
      />
    );
    const closeSyncView = () => {
      if (selectedDevice) {
        selectDevice(undefined);
        toggleDeviceSelect(selectedDevice);
      }
      const { listDevices } = this.props;
      listDevices();
      navigation.goBack();
    };
    const handleModalContinue = () => {
      hideSyncedModal();
    };

    return (
      <View
        style={{
          flex: 1,
          backgroundColor: MAPEO_BLUE
        }}
      >
        {noDevices ? (
          <View style={{ flex: 1 }}>
            <SyncHeader
              closeSyncView={closeSyncView}
              deviceText={headerDeviceText}
              syncStopped={syncStopped}
            />
            <View
              style={{
                flex: 1,
                alignItems: 'center',
                justifyContent: 'center'
              }}
            >
              <View
                style={{
                  width: 200,
                  height: 200,
                  backgroundColor: MEDIUM_BLUE,
                  borderRadius: 100,
                  alignItems: 'center',
                  justifyContent: 'center'
                }}
              >
                <WifiIcon name="wifi" size={180} color="white" />
              </View>
              <Text
                style={{
                  marginTop: 30,
                  color: 'white',
                  fontWeight: '700',
                  fontSize: 18,
                  textAlign: 'center'
                }}
              >
                {I18n.t('sync.check')}
              </Text>
            </View>
          </View>
        ) : (
          <View>
            <FlatList
              scrollEnabled
              ListHeaderComponent={
                <SyncHeader
                  closeSyncView={closeSyncView}
                  deviceText={headerDeviceText}
                  syncStopped={syncStopped}
                />
              }
              data={devices}
              keyExtractor={keyExtractor}
              renderItem={renderItem}
              style={{ width: Dimensions.get('window').width }}
            />
            <SyncedModal
              onContinue={handleModalContinue}
              visible={syncedModalVisible}
            />
          </View>
        )}
      </View>
    );
  }
}

export default withNavigationFocus(SyncView);
