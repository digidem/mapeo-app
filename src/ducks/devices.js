// @flow

import update from 'immutability-helper';
import { keyBy } from 'lodash';
import { create } from '../lib/redux';

export const {
  type: ANNOUNCE_SYNC,
  action: announceSync,
  reducer: announceSyncReducer
} = create('ANNOUNCE_SYNC', {
  success: (state, action) => {
    const newState = update(state, {
      $toggle: ['announceSync']
    });

    return newState;
  }
});

export const {
  type: DEVICE_LIST,
  action: deviceList,
  reducer: deviceListReducer
} = create('DEVICE_LIST', {
  success: (state, action) => {
    const newState = update(state, {
      devices: { $set: keyBy(action.payload, 'id') }
    });

    return newState;
  }
});

export const {
  type: DEVICE_TOGGLE_SELECT,
  action: deviceToggleSelect,
  reducer: deviceToggleSelectReducer
} = create('DEVICE_TOGGLE_SELECT', {
  start: (state, action) => {
    const newState = update(state, {
      devices: {
        [action.meta.id]: {
          $toggle: ['selected']
        }
      }
    });

    return newState;
  }
});

export const {
  type: DEVICE_SELECT,
  action: deviceSelect,
  reducer: deviceSelectReducer
} = create('DEVICE_SELECT', {
  start: (state, action) => {
    const newState = update(state, {
      selectedDevice: {
        $set: action.meta
      }
    });

    return newState;
  }
});

export const {
  type: DEVICE_SYNC_UPDATE,
  action: deviceSyncUpdate,
  reducer: deviceSyncUpdateReducer
} = create('DEVICE_SYNC_UPDATE', {
  start: (state, action) => {
    const newState = update(state, {
      devices: {
        [action.meta.id]: {
          $set: action.meta
        }
      }
    });

    return newState;
  }
});

export default [
  announceSyncReducer,
  deviceListReducer,
  deviceToggleSelectReducer,
  deviceSelectReducer,
  deviceSyncUpdateReducer
];
