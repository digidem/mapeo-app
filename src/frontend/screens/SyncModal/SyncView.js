// @flow
import React from "react";
import { View, ScrollView, Text, StyleSheet } from "react-native";
import { TouchableNativeFeedback } from "../../sharedComponents/Touchables";

import IconButton from "../../sharedComponents/IconButton";
import { CloseIcon, WifiOffIcon, WifiIcon } from "../../sharedComponents/icons";
import DotIndicator from "./DotIndicator";
import PeerList from "./PeerList";
import type { Peer } from "./PeerList";

type HeaderProps = {
  onClosePress: () => void
};

const Header = ({ onClosePress }: HeaderProps) => (
  <View style={styles.header}>
    <IconButton onPress={onClosePress}>
      <CloseIcon color="white" />
    </IconButton>
    <Text numberOfLines={1} style={styles.title}>
      Sincronizar
    </Text>
  </View>
);

const WifiBar = ({ onPress, ssid, deviceName }) => (
  <TouchableNativeFeedback onPress={onPress}>
    <View style={styles.wifiBar}>
      <WifiIcon />
      <Text style={styles.wifiBarText}>
        <Text style={styles.bold}>Wi-Fi:</Text> {ssid}, {deviceName}
      </Text>
    </View>
  </TouchableNativeFeedback>
);

const NoWifiBox = ({ onPress }) => (
  <TouchableNativeFeedback onPress={onPress}>
    <View style={styles.noWifiBox}>
      <View style={styles.noWifiBoxInner}>
        <View style={styles.noWifiIcon}>
          <WifiOffIcon size={50} />
        </View>
        <View style={styles.noWifiTextContainer}>
          <Text style={styles.infoHeader}>Sin Wi-Fi</Text>
          <Text style={styles.infoSubheader}>
            Abrir la configuración y conectar a una red Wi-Fi para sincronizar
          </Text>
        </View>
      </View>
    </View>
  </TouchableNativeFeedback>
);

const SearchingBox = () => (
  <View style={styles.searchingBox}>
    <View style={styles.searchingBoxInner}>
      <DotIndicator />
      <View style={styles.searchingTextContainer}>
        <Text style={styles.infoHeader}>Buscando</Text>
        <Text style={styles.infoSubheader}>
          Asegurar que los otros equipos están encendidos y conectados a la
          misma red de Wi-Fi
        </Text>
      </View>
    </View>
  </View>
);

type Props = {
  onClosePress: () => void,
  onSyncPress: (peerId: string) => void,
  onWifiPress: () => void,
  deviceName: string,
  peers: Array<Peer>,
  ssid: null | string
};

const SyncView = ({
  onClosePress,
  onSyncPress,
  peers,
  ssid,
  deviceName,
  onWifiPress
}: Props) => (
  <ScrollView
    style={styles.container}
    contentContainerStyle={styles.scrollViewContent}
  >
    <Header onClosePress={onClosePress} />
    {ssid ? (
      <>
        <WifiBar onPress={onWifiPress} ssid={ssid} deviceName={deviceName} />
        {peers.length ? (
          <PeerList peers={peers} onSyncPress={onSyncPress} />
        ) : (
          <SearchingBox />
        )}
      </>
    ) : (
      <NoWifiBox onPress={onWifiPress} />
    )}
  </ScrollView>
);

export default SyncView;

const styles = StyleSheet.create({
  header: {
    flexGrow: 0,
    flexShrink: 0,
    height: 60,
    flexDirection: "row",
    alignItems: "center",
    backgroundColor: "#3366FF"
  },
  title: {
    color: "white",
    fontWeight: "700",
    fontSize: 20,
    flex: 1,
    textAlign: "center",
    marginRight: 60
  },
  container: {
    flex: 1,
    backgroundColor: "#2348B2"
  },
  scrollViewContent: {
    flex: 1,
    flexDirection: "column",
    justifyContent: "flex-start"
  },
  noWifiBox: {
    backgroundColor: "#AD145C",
    minHeight: 300,
    paddingHorizontal: 30,
    flexDirection: "column",
    justifyContent: "center"
  },
  noWifiBoxInner: {
    flex: 1,
    flexDirection: "row",
    alignItems: "center"
  },
  searchingBoxInner: {
    flex: 1,
    flexDirection: "row",
    alignItems: "center"
  },
  noWifiIcon: {
    width: 120,
    height: 120,
    backgroundColor: "#93114E",
    borderRadius: 60,
    alignItems: "center",
    justifyContent: "center"
  },
  infoHeader: {
    color: "white",
    fontWeight: "700",
    fontSize: 24
  },
  infoSubheader: {
    color: "white",
    fontWeight: "400",
    fontSize: 20
  },
  noWifiTextContainer: {
    maxWidth: "60%",
    marginLeft: 20
  },
  searchingTextContainer: {
    maxWidth: "75%",
    marginLeft: 30
  },
  wifiBar: {
    backgroundColor: "#19337F",
    height: 50,
    flexDirection: "row",
    alignItems: "center",
    justifyContent: "flex-start",
    paddingHorizontal: 15
  },
  wifiBarText: {
    color: "white",
    paddingLeft: 10
  },
  bold: {
    fontWeight: "700"
  },
  searchingBox: {
    backgroundColor: "#2348B2",
    minHeight: 250,
    paddingHorizontal: 30,
    flexDirection: "column",
    justifyContent: "center",
    alignItems: "stretch"
  }
});
