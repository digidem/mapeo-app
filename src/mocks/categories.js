// @flow

export const createCategory = (category?: any) => ({
  id: 'categoryID',
  icon: require('../images/categories/category_18.png'),
  name: 'categoryName',
  fields: [],

  ...category
});
