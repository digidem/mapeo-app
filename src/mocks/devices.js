// @flow

export const createDevice = (device?: any) => ({
  id: '0',
  name: 'deviceName',
  selected: false,
  syncStatus: null,

  ...device
});
