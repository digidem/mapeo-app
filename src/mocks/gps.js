// @flow
import type { GPSState } from '../types/gps';
import {
  PERMISSION_DENIED,
  UNAVAILABLE,
  SEARCHING,
  LOW_ACCURACY,
  HIGH_ACCURACY
} from '../ducks/gps';

export const createGPSState = (partial: any): GPSState => ({
  status: HIGH_ACCURACY,
  coords: {
    longitude: 1,
    latitude: 1,
    accuracy: 50
  },
  timestamp: 0,

  ...partial
});
