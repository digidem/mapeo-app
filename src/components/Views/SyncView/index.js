// @flow

import { connect } from 'react-redux';
import { values } from 'lodash';
import type { Dispatch } from 'redux';
import { StoreState } from '../../../types/redux';
import {
  syncAnnounce,
  syncStart,
  deviceSelect,
  deviceToggleSelect,
  deviceSyncUpdate
} from '../../../ducks/devices';
import { modalShow, modalHide } from '../../../ducks/modals';
import SyncView from './SyncView';
import type { StateProps, DispatchProps } from './SyncView';

function mapStateToProps(state: StoreState): StateProps {
  return {
    devices: values(state.app.devices),
    selectedDevice: state.app.selectedDevice,
    syncedModalVisible: state.app.modals.synced
  };
}

function mapDispatchToProps(dispatch: Dispatch): DispatchProps {
  return {
    announceSync: () => dispatch(syncAnnounce()),
    startSync: device => dispatch(syncStart(device)),
    selectDevice: device => dispatch(deviceSelect(device)),
    toggleDeviceSelect: device => dispatch(deviceToggleSelect(device)),
    updateDeviceSync: device => {
      dispatch(deviceSelect(device));
      dispatch(deviceSyncUpdate(device));
    },
    showSyncedModal: () => dispatch(modalShow('synced')),
    hideSyncedModal: () => dispatch(modalHide('synced'))
  };
}

export default connect(mapStateToProps, mapDispatchToProps)(SyncView);
