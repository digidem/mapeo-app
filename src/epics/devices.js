// @flow

import React from 'react';
import type { ActionsObservable } from 'redux-observable';
import {
  SYNC_ANNOUNCE,
  SYNC_START,
  syncStart,
  DEVICE_LIST,
  deviceList
} from '../ducks/devices';
import { createDevice } from '../mocks/devices';
import type { Action } from '../types/redux';
import type { Device } from '../types/device';
import type { StoreState } from '../types/redux';
import Sync from '../api/sync';

const initialDevices = [
  {
    name: `Aldo's MacBook`,
    host: '192.168.0.1',
    port: 123,
    selected: false,
    syncStatus: 'notStarted'
  },
  {
    name: `Cindy's MacBook`,
    host: '192.168.0.1',
    port: 123,
    selected: false,
    syncStatus: 'notStarted'
  },
  {
    name: `Stephen's Laptop`,
    host: '192.168.0.1',
    port: 123,
    selected: false,
    syncStatus: 'notStarted'
  },
  {
    name: `Karissa's MacBook`,
    host: '192.168.0.1',
    port: 123,
    selected: false,
    syncStatus: 'notStarted'
  },
  {
    name: `Gregor's MacBook`,
    host: '192.168.0.1',
    port: 123,
    selected: false,
    syncStatus: 'notStarted'
  }
];

// const initialDevices = [];

export const syncAnnounceEpic = (action$: ActionsObservable<any>) =>
  action$
    .ofType(SYNC_ANNOUNCE)
    .filter(action => action.status === 'Start')
    .flatMap(() => Sync.announce().map(response => deviceList('')));

export const syncStartEpic = (action$: ActionsObservable<any>) =>
  action$
    .ofType(SYNC_START)
    .filter(action => action.status === 'Start')
    .flatMap(action =>
      Sync.start(action.meta).map(response => {
        return syncStart(action.meta, response);
      })
    );

export const deviceListEpic = (
  action$: ActionsObservable<Action<string, Device[]>>,
  store: StoreState
) =>
  action$
    .ofType(DEVICE_LIST)
    .filter(action => action.status === 'Start')
    .flatMap(() =>
      Sync.list().map(devices =>
        deviceList(
          '',
          devices.map((d, i) =>
            createDevice({
              name: d.name,
              id: i.toString(),
              host: d.host,
              port: d.port,
              selected: false,
              syncStatus: d.status,
              message: d.message
            })
          )
        )
      )
    );

export default [syncAnnounceEpic, syncStartEpic, deviceListEpic];
