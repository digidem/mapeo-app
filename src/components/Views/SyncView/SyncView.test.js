// @flow
import React from 'react';
import 'react-native';
import renderer from 'react-test-renderer';
import SyncView from './SyncView';
import { createDevice } from '../../../mocks/devices';

describe('SyncView tests', () => {
  const device = createDevice();
  const cases = [device];
  const isFocused = () => true;
  const addListener = () => true;
  const announceSync = jest.fn();
  const listDevices = jest.fn();
  jest.mock('./DeviceCell', () => 'mock-device-cell');

  test('snapshot', () => {
    const tree = renderer.create(
      <SyncView
        announceSync={announceSync}
        devices={cases}
        navigation={{ addListener, isFocused }}
        listDevices={listDevices}
      />
    );
    expect(tree).toMatchSnapshot();
  });
});
