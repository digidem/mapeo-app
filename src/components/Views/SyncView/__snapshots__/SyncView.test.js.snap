// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SyncView tests snapshot 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#3366ff",
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <View
      style={
        Object {
          "flexDirection": "column",
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#3366ff",
            "flexDirection": "row",
            "height": 65,
            "justifyContent": "center",
            "paddingHorizontal": 15,
          }
        }
      >
        <View
          accessible={true}
          isTVSelectable={true}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "backgroundColor": "transparent",
              "left": 15,
              "opacity": 1,
              "position": "absolute",
            }
          }
        >
          <Text
            accessible={true}
            allowFontScaling={false}
            ellipsizeMode="tail"
            style={
              Array [
                Object {
                  "color": "#ffffff",
                  "fontSize": 30,
                },
                undefined,
                Object {
                  "fontFamily": "Material Design Icons",
                  "fontStyle": "normal",
                  "fontWeight": "normal",
                },
              ]
            }
          >
            
          </Text>
        </View>
        <View
          style={
            Object {
              "flexDirection": "column",
            }
          }
        >
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
            style={
              Object {
                "alignSelf": "flex-start",
                "color": "#ffffff",
                "fontSize": 18,
                "fontWeight": "700",
              }
            }
          >
            Sync
          </Text>
        </View>
      </View>
      <View
        style={
          Array [
            Object {
              "backgroundColor": "#17317c",
              "flexDirection": "row",
              "justifyContent": "space-between",
              "paddingHorizontal": 10,
              "paddingVertical": 10,
            },
            null,
          ]
        }
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Object {
              "color": "#ffffff",
              "fontSize": 12,
              "fontWeight": "700",
            }
          }
        >
          No Available Devices
        </Text>
      </View>
    </View>
    <View
      style={
        Object {
          "alignItems": "center",
          "flex": 1,
          "justifyContent": "center",
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#2752d1",
            "borderRadius": 100,
            "height": 200,
            "justifyContent": "center",
            "width": 200,
          }
        }
      >
        <Text
          accessible={true}
          allowFontScaling={false}
          ellipsizeMode="tail"
          style={
            Array [
              Object {
                "color": "white",
                "fontSize": 180,
              },
              undefined,
              Object {
                "fontFamily": "Material Icons",
                "fontStyle": "normal",
                "fontWeight": "normal",
              },
            ]
          }
        >
          
        </Text>
      </View>
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "color": "white",
            "fontSize": 18,
            "fontWeight": "700",
            "marginTop": 30,
            "textAlign": "center",
          }
        }
      >
        Make sure devices are turned on and connected to WiFi
      </Text>
    </View>
  </View>
</View>
`;
